crates/bitwarden-ipc/src/ipc_client.rs

```diff
@@ -1,233 +1,288 @@
-where
-    Crypto: CryptoProvider<Com, Ses>,
-    Com: CommunicationBackend,
-    Ses: SessionRepository<Session = Crypto::Session>,
-    Payload: TryFrom<Vec<u8>> + PayloadTypeName,
-{
-    receiver: Com::Receiver,
-    client: Arc<IpcClient<Crypto, Com, Ses>>,
-    _payload: std::marker::PhantomData<Payload>,
+pub struct IpcClientTypedSubscription<Payload: TryFrom<Vec<u8>> + PayloadTypeName>(
+    IpcClientSubscription,
+    std::marker::PhantomData<Payload>,
+);
+
+#[derive(Debug, Error, Clone, PartialEq, Eq)]
+#[bitwarden_error(flat)]
+pub enum SubscribeError {
+    #[error("The IPC processing thread is not running")]
+    NotStarted,
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+#[bitwarden_error(flat)]
+pub enum ReceiveError {
+    #[error("Failed to subscribe to the IPC channel: {0}")]
+    Channel(#[from] tokio::sync::broadcast::error::RecvError),
+
+    #[error("Timed out while waiting for a message: {0}")]
+    Timeout(#[from] tokio::time::error::Elapsed),
+
+    #[error("Cancelled while waiting for a message")]
+    Cancelled,
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+#[bitwarden_error(flat)]
+pub enum TypedReceiveError {
+    #[error("Failed to subscribe to the IPC channel: {0}")]
+    Channel(#[from] tokio::sync::broadcast::error::RecvError),
+
+    #[error("Timed out while waiting for a message: {0}")]
+    Timeout(#[from] tokio::time::error::Elapsed),
+
+    #[error("Cancelled while waiting for a message")]
+    Cancelled,
+
+    #[error("Typing error: {0}")]
+    Typing(String),
+}
+
+impl From<ReceiveError> for TypedReceiveError {
+    fn from(value: ReceiveError) -> Self {
+        match value {
+            ReceiveError::Channel(e) => TypedReceiveError::Channel(e),
+            ReceiveError::Timeout(e) => TypedReceiveError::Timeout(e),
+            ReceiveError::Cancelled => TypedReceiveError::Cancelled,
+        }
+    }
 }
 
 impl<Crypto, Com, Ses> IpcClient<Crypto, Com, Ses>
 where
     Crypto: CryptoProvider<Com, Ses>,
     Com: CommunicationBackend,
-    Ses: SessionRepository<Session = Crypto::Session>,
+    Ses: SessionRepository<Crypto::Session>,
 {
     pub fn new(crypto: Crypto, communication: Com, sessions: Ses) -> Arc<Self> {
         Arc::new(Self {
             crypto,
             communication,
             sessions,
+
+            incoming: RwLock::new(None),
+            cancellation_token: RwLock::new(None),
         })
     }
 
+    pub async fn start(self: &Arc<Self>) {
+        let cancellation_token = CancellationToken::new();
+        let mut self_cancellation_token = self.cancellation_token.write().await;
+
+        *self_cancellation_token = Some(cancellation_token.clone());
+
+        let com_receiver = self.communication.subscribe().await;
+        let (client_tx, client_rx) = tokio::sync::broadcast::channel(CHANNEL_BUFFER_CAPACITY);
+
+        let mut client_incoming = self.incoming.write().await;
+        *client_incoming = Some(client_rx);
+        let _ = client_incoming;
```

Comment:
Here we're dropping correctly, but I feel like using replace makes it easier to parse.
```suggestion
        self.incoming.write().await.replace(client_rx);
```

Reply:
Fixed!